lib/nerve.rb

```diff
@@ -86,15 +165,47 @@ def run
       $EXIT = true
     end
 
-    def launch_watcher(name, config)
-      log.debug "nerve: launching service watcher #{name}"
-      watcher = ServiceWatcher.new(config.merge({'instance_id' => @instance_id, 'name' => name}))
-      @watchers[name] = Thread.new{watcher.run}
+    def merged_config(config, name)
+      # Get a deep copy so sub-hashes are properly handled
+      deep_copy = Marshal.load(Marshal.dump(config))
+      return deep_copy.merge({'instance_id' => @instance_id, 'name' => name})
+    end
+
+    def launch_watcher(name, config, wait=false)
+      watcher_config = merged_config(config, name)
+      # The ServiceWatcher may mutate the configs, so record the version before
+      # passing the config to the ServiceWatcher
+      @watcher_versions[name] = watcher_config.hash
+
+      watcher = ServiceWatcher.new(watcher_config)
+      unless @config_manager.options[:check_config]
+        log.debug "nerve: launching service watcher #{name}"
+        watcher_thread = Thread.new { watcher.run }
+        @watchers[name] = watcher_thread
+        if wait
+          log.info "nerve: waiting for watcher thread #{name} to report"
+          responsive_sleep(30) { break if watcher_thread[:has_reported] }
+          log.info "nerve: watcher thread #{name} has reported!"
+        end
+      else
+        log.info "nerve: not launching #{name} due to --check-config option"
+      end
     end
 
     def reap_watcher(name)
       watcher_thread = @watchers.delete(name)
-      watcher_thread.join()
+      @watcher_versions.delete(name)
+      # Signal the watcher thread to exit
+      watcher_thread[:finish] = true
+
+      unclean_shutdown = watcher_thread.join(10).nil?
```

Comment:
this functionality should really be delegated to the watcher itself. it should be sufficient to just call `watcher.stop` and know that it's going to work.